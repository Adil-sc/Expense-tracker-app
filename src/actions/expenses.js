import uuid from 'uuid';
import database from '../firebase/firebase';

// Actions:
// They are JS objects, with a type property that descibes the action + an optional payload.
// Actions are displatched using store.dispatch. Dispatching means to send an action to the Reducer.

// Add Expense action
export const addExpense = (expense) =>{
    return {
        type: 'ADD_EXPENSE',
        expense:expense
    }
}

export const startAddExpense = (expenseData = {})=>{
    // Thunked/async actions get called with dispatch and getState
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        const {
            description='',
            note='',
            amount=0,
            createdAt=0
        } = expenseData //destructure from expense data, just another way of doing it

        const expense = {description, note, amount, createdAt}
        
        // Returning so we can make use of promise chaning 
        // We push the data to firebase and then call the dispatch action for adding the data to the redux store
       return database.ref(`users/${uid}/expenses`).push(expense).then((ref)=>{
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    }
}

// REMOVE_EXPENSE
export const removeExpense = ({id}={})=>{
    return {
        type:'REMOVE_EXPENSE',
        id:id
    }
};

export const startRemoveExpense =({id}={}) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        // Remove the epxense with the specific id from Firebase DB
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            // Make sure its also remvoed from the redux store
            dispatch(removeExpense({id:id}));
        });
    };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) =>{
    return {
        type:'EDIT_EXPENSE',
        id:id,
        updates:updates     //An object of expense properties we want to update
    }
}

export const startEditExpense = (id, updates) =>{
    
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update({
            ...updates
        }).then(()=>{
            dispatch(editExpense(id, updates));
        });
    }
}


// SET_EXPENSES - aka fetch the data from Firebase DB
export const setExpenses = (expenses) =>{
    return {
        type:'SET_EXPENSES',
        expenses: expenses
    }
};

export const startSetExpenses = () =>{
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        // Fetch the data from the Firebase DB and add push it into the expenses array
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot)=>{
            const expenses = [];

            // Convert object structure from Firebase, into an array structure
            snapshot.forEach((childSnapshot)=>{
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            // Call the setExpenses action and pass the expenses array to the reducer, so that the
            // reducer can update the state to include all expenses that live on the Firebase DB
            dispatch(setExpenses(expenses));

        });

    }
};
