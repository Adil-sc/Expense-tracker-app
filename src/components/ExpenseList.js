import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';


export const ExpenseList = (props) =>{
    return (
        <div>
           {
               props.expenses.length === 0 ? (
                <p>No expenses</p>
               ):(
                props.expenses.map((expense)=>{
                    return <ExpenseListItem key={expense.id} {...expense} />
                })      
               )
           }
            
        </div>
    );
};


const mapStateToProps = (state)=>{
    // This object represents the diff properties we want to add to the prop
    return {
        expenses:selectExpenses(state.expenses, state.filters),
        filters:state.filters
    };
};


// mapStateToProps defines what we want to grab off the Store, and ExpenseList is componenet we want to create the connected version of
// Result is a brand new componenet, ExpenseList + props from the Store.
// Connect() returns the higher order compoenet, which then wraps the ExpenseList component
export default connect(mapStateToProps)(ExpenseList);