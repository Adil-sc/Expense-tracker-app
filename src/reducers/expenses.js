
// Expenses reducer (The reducer is a pure function that takes the previous state and an action, and returns the next state)
const expensesReducerDefaultState = []
const expensesReducer = (state=expensesReducerDefaultState, action) =>{
    switch(action.type){
        case 'ADD_EXPENSE':
            //...state is a spread operator, inserts all values in the state array
            // When we return, this will be the new state object, so we need to make sure we don't override the whole state object
            // That is why we use ...state and append action.expense property onto it
            return [...state,action.expense]; 
        case 'REMOVE_EXPENSE':
            return state.filter((expense)=>{
                return expense.id !== action.id
            });
        case 'EDIT_EXPENSE':
            return state.map((expense)=>{
                if (expense.id === action.id){
                    return {
                        ...expense,           //Object spread operator, makes a copy of expense object
                        ...action.updates     //Object spread operator, we update the properties in the expense object copy (ex amount)
                    }
                }else{
                    return expense
                }
            });
        default:
            return state;
    }
};


export default expensesReducer;