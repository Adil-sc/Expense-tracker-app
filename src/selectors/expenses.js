// A "Selector" is a function which accepts a Redux state as an argument, and then returns
// data that is derived from that state (aka pick out a value from the state). They help to keep the redux store lean and bloat free

// Get visible expenses - returns an array of filtered objects
import moment from "moment";

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};

export default getVisibleExpenses;
