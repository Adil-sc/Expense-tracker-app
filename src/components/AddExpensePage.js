import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';

 
export class AddExpensePage extends React.Component{
    onSubmit = (expense)=>{
        // Add expense data to redux store
        // props.dispatch(addExpense(expense));
        this.props.addExpense(expense);
        // Automatically redirect to dashboard page after creating a new expense
        this.props.history.push('/');
    }

    render(){
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm 
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}




const mapDispatchToProps = (dispatch) =>{
    return {
        addExpense: (expense) => {
            return dispatch(addExpense(expense));
        }
    };
};


export default connect(undefined, mapDispatchToProps )(AddExpensePage);