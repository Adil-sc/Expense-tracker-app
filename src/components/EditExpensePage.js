import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, startRemoveExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component {

    onSubmit = (expense) =>{
        this.props.editExpense(this.props.expense.id, expense);
        // console.log('updated', expense);
        this.props.history.push('/');
    }

    onRemove = () =>{
        this.props.startRemoveExpense({id:this.props.expense.id});
        this.props.history.push('/');
    }

    render(){
        return (
            <div>
                 <ExpenseForm
                     expense={this.props.expense} 
                     onSubmit={this.onSubmit}
                 />
                 <button onClick={this.onRemove}>Remove</button>
            </div>
         );
    }

}



const mapStateToProps = (state, props) =>{
    return {
        // props.match.params is a key/value pair object passed from the URL. We check to see if 
        // any expenses in our store match the expense with the edit/id in the URL.If so, we return it
        // The props object should then contain the matched expense object
        expense: state.expenses.find((expense)=>{
            return expense.id === props.match.params.id
        })
    }
}

const mapDispatchToProps=(dispatch,props)=>{
    return{
        editExpense: (id, expense)=>{
            return dispatch(editExpense(id,expense));
        },
        startRemoveExpense:(data)=>{
            return dispatch(startRemoveExpense(data));
        }
    }
}

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);