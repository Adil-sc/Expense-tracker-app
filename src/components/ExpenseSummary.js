import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import numeral from "numeral";
import getVisibleExpenses from "../selectors/expenses";
import { getExpensesTotal } from "../selectors/expense-total";

export const ExpenseSummary = (props) => {
  return (
    <div className="page-header">
      <div className="content-container">
        <h2 className="page-header__title">
          {" "}
          <span>{props.expenseCount}</span> expenses totalling{" "}
          <span>{numeral(props.expensesTotal / 100).format("$0,0.00")}</span>{" "}
        </h2>
        <div className="page-header__actions">
          <Link className="button" to="/create">
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenseCount: getVisibleExpenses(state.expenses, state.filters).length,
    expensesTotal: getExpensesTotal(
      getVisibleExpenses(state.expenses, state.filters)
    ),
  };
};

export default connect(mapStateToProps)(ExpenseSummary);
