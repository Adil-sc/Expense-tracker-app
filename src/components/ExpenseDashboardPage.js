import React from 'react';
import ExpenseList from './ExpenseList.js';
import ExpenseListFilters from '../components/ExpenseListFilters.js';
import ExpenseSummary from '../components/ExpenseSummary';

const ExpenseDashboardPage = () => {
    return  (
        <div>
            <ExpenseSummary />
            <ExpenseListFilters />
            <ExpenseList />
        </div>
    );
};

export default ExpenseDashboardPage;