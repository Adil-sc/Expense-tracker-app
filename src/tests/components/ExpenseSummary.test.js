import React from 'react'
import {shallow} from 'enzyme'
import {expenses} from '../fixtures/expenses';
import {ExpenseSummary} from '../../components/ExpenseSummary';

test('should render ExpenseSummary component', ()=>{
    const wrapper = shallow(<ExpenseSummary />);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpenseSummary with 1 expense',()=>{
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expensesTotal={235} />);
    expect(wrapper).toMatchSnapshot();

});


test('should correctly render ExpenseSummary with multiple expense',()=>{
    const wrapper = shallow(<ExpenseSummary expenseCount={10} expensesTotal={53433435} />);
    expect(wrapper).toMatchSnapshot();

});