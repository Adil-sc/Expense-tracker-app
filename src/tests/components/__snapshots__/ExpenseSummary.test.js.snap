// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpenseSummary with 1 expense 1`] = `
<div>
  <h2>
     
    1
     expenses totalling 
    $2.35
     
  </h2>
</div>
`;

exports[`should correctly render ExpenseSummary with multiple expense 1`] = `
<div>
  <h2>
     
    10
     expenses totalling 
    $534,334.35
     
  </h2>
</div>
`;

exports[`should render ExpenseSummary component 1`] = `
<div>
  <h2>
     
     expenses totalling 
    $0.00
     
  </h2>
</div>
`;
